## Data Preprocessing:

In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [22]:
import warnings
warnings.filterwarnings("ignore")

In [23]:
df = pd.read_excel('/content/sample_data/Online retail.xlsx',header =None,names = ['Transaction'])

In [24]:
df.head()

Transaction
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [26]:
transactions = df['Transaction'].apply(lambda x:x.split(','))

In [27]:
#df.head()

In [28]:
# Create a list of unique items
items = sorted(set(item for transaction in transactions for item in transaction))


In [29]:
basket = pd.DataFrame(0, index=range(len(transactions)), columns=items)
for i, transaction in enumerate(transactions):
    for item in transaction:
        basket.at[i, item] = 1


In [30]:
basket.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        1                  1          0        1            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        0            0   
3           0        0                  0          0        1            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               1   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       0               0   
3      0               0          0            0  ...       1               0   
4      0               0          0            0  ...       0               0   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0            0           0                 1                  0   
1            0           0                 0                  0   
2            0           0                 0                  0   
3            0           0                 0                  0   
4            0           0                 0                  0   

   whole wheat rice  yams  yogurt cake  zucchini  
0                 0     1            0         0  
1                 0     0            0         0  
2                 0     0            0         0  
3                 0     0            0         0  
4                 1     0            0         0  

[5 rows x 120 columns]

In [31]:
basket.shape

(7501, 120)

## Association Rule Mining:

In [35]:
from mlxtend.frequent_patterns import apriori, association_rules

In [36]:
frequent_itemsets = apriori(basket,min_support = 0.005, use_colnames=True)
rules = association_rules(frequent_itemsets,metric='lift',min_threshold=1)
rules

antecedents             consequents  antecedent support  \
0                  (almonds)               (burgers)            0.020397   
1                  (burgers)               (almonds)            0.087188   
2                  (almonds)             (chocolate)            0.020397   
3                (chocolate)               (almonds)            0.163845   
4                  (almonds)                  (eggs)            0.020397   
...                      ...                     ...                 ...   
1827  (olive oil, spaghetti)              (pancakes)            0.022930   
1828   (pancakes, spaghetti)             (olive oil)            0.025197   
1829             (olive oil)   (pancakes, spaghetti)            0.065858   
1830              (pancakes)  (olive oil, spaghetti)            0.095054   
1831             (spaghetti)   (olive oil, pancakes)            0.174110   

      consequent support   support  confidence      lift  leverage  \
0               0.087188  0.005199    0.254902  2.923577  0.003421   
1               0.020397  0.005199    0.059633  2.923577  0.003421   
2               0.163845  0.005999    0.294118  1.795099  0.002657   
3               0.020397  0.005999    0.036615  1.795099  0.002657   
4               0.179709  0.006532    0.320261  1.782108  0.002867   
...                  ...       ...         ...       ...       ...   
1827            0.095054  0.005066    0.220930  2.324260  0.002886   
1828            0.065858  0.005066    0.201058  3.052910  0.003407   
1829            0.025197  0.005066    0.076923  3.052910  0.003407   
1830            0.022930  0.005066    0.053296  2.324260  0.002886   
1831            0.010799  0.005066    0.029096  2.694478  0.003186   

      conviction  zhangs_metric  
0       1.225089       0.671653  
1       1.041724       0.720799  
2       1.184553       0.452150  
3       1.016834       0.529719  
4       1.206774       0.448005  
...          ...            ...  
1827    1.161572       0.583127  
1828    1.169224       0.689825  
1829    1.056037       0.719852  
1830    1.032075       0.629602  
1831    1.018846       0.761446  

[1832 rows x 10 columns]

In [37]:
frequent_itemsets = apriori(basket,min_support = 0.01, use_colnames=True)
rules = association_rules(frequent_itemsets,metric='lift',min_threshold=1)
rules

antecedents                 consequents  \
0                     (avocado)             (mineral water)   
1               (mineral water)                   (avocado)   
2                        (cake)                   (burgers)   
3                     (burgers)                      (cake)   
4                     (burgers)                 (chocolate)   
..                          ...                         ...   
401  (mineral water, spaghetti)                  (pancakes)   
402       (pancakes, spaghetti)             (mineral water)   
403             (mineral water)       (pancakes, spaghetti)   
404                  (pancakes)  (mineral water, spaghetti)   
405                 (spaghetti)   (mineral water, pancakes)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.033329            0.238368  0.011598    0.348000  1.459926   
1              0.238368            0.033329  0.011598    0.048658  1.459926   
2              0.081056            0.087188  0.011465    0.141447  1.622319   
3              0.087188            0.081056  0.011465    0.131498  1.622319   
4              0.087188            0.163845  0.017064    0.195719  1.194537   
..                  ...                 ...       ...         ...       ...   
401            0.059725            0.095054  0.011465    0.191964  2.019529   
402            0.025197            0.238368  0.011465    0.455026  1.908923   
403            0.238368            0.025197  0.011465    0.048098  1.908923   
404            0.095054            0.059725  0.011465    0.120617  2.019529   
405            0.174110            0.033729  0.011465    0.065850  1.952333   

     leverage  conviction  zhangs_metric  
0    0.003654    1.168147       0.325896  
1    0.003654    1.016113       0.413630  
2    0.004398    1.063198       0.417434  
3    0.004398    1.058080       0.420238  
4    0.002779    1.039630       0.178411  
..        ...         ...            ...  
401  0.005788    1.119933       0.536902  
402  0.005459    1.397557       0.488452  
403  0.005459    1.024059       0.625163  
404  0.005788    1.069244       0.557862  
405  0.005593    1.034385       0.590626  

[406 rows x 10 columns]

In [51]:
rules['antecedents'].value_counts()

antecedents
(mineral water)                       42
(spaghetti)                           36
(chocolate)                           31
(milk)                                27
(eggs)                                24
(ground beef)                         18
(frozen vegetables)                   17
(green tea)                           16
(french fries)                        15
(pancakes)                            14
(burgers)                             12
(cake)                                10
(shrimp)                              10
(olive oil)                            9
(tomatoes)                             9
(chicken)                              7
(low fat yogurt)                       7
(mineral water, spaghetti)             7
(turkey)                               7
(whole wheat rice)                     6
(frozen smoothie)                      6
(grated cheese)                        5
(mineral water, milk)                  5
(cooking oil)                          5
(soup)                                 4
(herb & pepper)                        4
(mineral water, ground beef)           4
(mineral water, eggs)                  4
(mineral water, chocolate)             4
(salmon)                               3
(escalope)                             3
(spaghetti, chocolate)                 3
(eggs, chocolate)                      2
(milk, spaghetti)                      2
(eggs, spaghetti)                      2
(frozen vegetables, mineral water)     2
(milk, chocolate)                      2
(honey)                                2
(red wine)                             2
(frozen vegetables, spaghetti)         1
(milk, ground beef)                    1
(avocado)                              1
(spaghetti, ground beef)               1
(olive oil, mineral water)             1
(olive oil, spaghetti)                 1
(mineral water, pancakes)              1
(frozen vegetables, milk)              1
(cookies)                              1
(mineral water, french fries)          1
(spaghetti, french fries)              1
(eggs, milk)                           1
(eggs, ground beef)                    1
(ground beef, chocolate)               1
(fresh bread)                          1
(champagne)                            1
(cereals)                              1
(pancakes, spaghetti)                  1
Name: count, dtype: int64

In [49]:
rules[ (rules['lift'] >= 3) &
      (rules['confidence'] >= 0.3) ]

antecedents    consequents  antecedent support  consequent support  \
214  (herb & pepper)  (ground beef)             0.04946            0.098254   

      support  confidence      lift  leverage  conviction  zhangs_metric  
214  0.015998     0.32345  3.291994  0.011138     1.33286        0.73246

## Analysis and Interpretation:

Customers who buy **herb & pepper** are **3.29 times** more likely to also buy **ground beef** (32.3% of the time). This suggests these items are commonly used together in recipes.

 Insights:
- **Product Placement**: Place these items near each other or bundle them in promotions.
- **Targeted Marketing**: Offer deals on ground beef to customers purchasing herb & pepper.

This strong association can help optimize sales strategies and product layout.